include/init.h

```diff
@@ -25,49 +26,98 @@ extern "C" {
 #define _SYS_INIT_LEVEL_POST_KERNEL	2
 #define _SYS_INIT_LEVEL_APPLICATION	3
 
+struct device;
+
+/**
+ * @brief Static init entry structure for each device driver or services
+ *
+ * @param dev pointer to a device driver instance structure. Can be NULL
+ * if this init entry is not used for a device driver.
+ * @param init init function for the init entry. Also used to initialize
+ * the device driver instance represented by
+ */
+struct init_entry {
+	int (*init)(struct device *dev);
+	struct device *dev;
+};
+
+void z_sys_init_run_level(s32_t level);
+
 /* A counter is used to avoid issues when two or more system devices
  * are declared in the same C file with the same init function.
  */
 #define Z_SYS_NAME(init_fn) _CONCAT(_CONCAT(sys_init_, init_fn), __COUNTER__)
 
 /**
- * @def SYS_INIT
+ * @def INIT_ENTRY_DEFINE
  *
- * @brief Run an initialization function at boot at specified priority
+ * @brief Create an init entry object and set it up for boot time initialization
  *
- * @details This macro lets you run a function at system boot.
+ * @details This macro defines an init entry object that will be automatically
+ * configured by the kernel during system initialization. Note that
+ * init entries will not be accessible from user mode. Also this macro should
+ * not be used directly, use relevant macro such as SYS_INIT() or
+ * DEVICE_AND_API_INIT() instead.
  *
- * @param init_fn Pointer to the boot function to run
+ * @param entry_name Init entry name. It is the name this instance exposes to
+ * the system.
  *
- * @param level The initialization level, See DEVICE_AND_API_INIT for details.
+ * @param init_fn Address to the init function of the entry.
  *
- * @param prio Priority within the selected initialization level. See
- * DEVICE_AND_API_INIT for details.
+ * @param device A device driver instance pointer or NULL
+ *
+ * @param level The initialization level at which configuration occurs.
+ * Must be one of the following symbols, which are listed in the order
+ * they are performed by the kernel:
+ * \n
+ * \li PRE_KERNEL_1: Used for devices that have no dependencies, such as those
```

Comment:
I am still not sure why PRE_KERNEL_1 and PRE_KERNEL_2 are meaningful. They used to be necessary, but now that we have proper dependency tracking in devicetree they do not seem needed. Further, their definitions here are about dependencies, but they don't really seem meaningful to me, since dependencies are a bit nuanced even with in the SoC: e.g. peripheral devices can depend on clock devices, etc.

Other than keeping legacy code working (which is not an objective of this effort), is there a reason to keep these levels? They seem to me to do more harm than good, but I don't know all the use cases, of course.

Reply:
This PR is not about changing the levels, so no harm here.

Anyway, note that each level have priority only between 0 to 99. That limits a bit the sorting possibility (devices ordering is made on level, then priority and finally their name). That was basically the reason of splitting the pre kernel level in 2, historically, if I remember well.

Reply:
> This PR is not about changing the levels, so no harm here.

Are you open to the discussion of removing those levels later then? I ask because this PR seems to be specifically about refactoring initialization, and that's what the levels are part of.

Reply:
No it's refactoring device structures only. Not levels, not the initialization (it's exactly the same process, I did not change anything there).

I would not change anything about the level yet. For the good reason that it is not yet a major issue, compared to the rest. level/prio dts generation is imo of a lesser priority compared to dts device instance generation (because if that's made right, then level/prio generation will be super trivial to do)

Let's proceed step by step (and make sure people go along each of these changes)

Reply:
> Are you open to the discussion of removing those levels later then?

Why do you want to remove them? those are not exclusive to device initialization or setup. A basic feature those levels provide is giving the developer the possibility to run code at any step of the booting process, sometimes you have to do HW initialization that is not managed by a driver or DTS and you need that at early stage. Removing those level will break many application and a useful interface.

Driver initialization priorities need to be improved and should utilise DTS for that, however, this is orthogonal and unrelated IMO. The driver priority happens on top of the existing scheme, does not replace it.

Reply:
Can you point to a definition for _1 vs. _2 that isn't in terms of device dependencies? Right now that's how they're defined. That's a hack. We have real dependencies from DT now.

Reply:
This is just a bug in the documentation, nobody bothered to change this and was long there and before device tree. So the fact it talks about devices does not mean this is 1:1 match with devicetree. Right now this is used for drivers, services and hardware initialization of different sorts that is not managed by DT, I am sure some of this can be optimized to be devicetree "devices" but we should not limit this to device tree, there is a need for this in the tree currently and there was always a need, the reason it is there.

I suggest the following change...

```diff
--- a/include/device.h
+++ b/include/device.h
@@ -57,25 +57,31 @@ extern "C" {
  * Must be one of the following symbols, which are listed in the order
  * they are performed by the kernel:
  * \n
- * \li PRE_KERNEL_1: Used for devices that have no dependencies, such as those
- * that rely solely on hardware present in the processor/SOC. These devices
+ * \li PRE_KERNEL_1: Used for units that have no dependencies, such as those
+ * that rely solely on hardware present in the processor/SOC. These units
  * cannot use any kernel services during configuration, since they are not
  * yet available.
  * \n
- * \li PRE_KERNEL_2: Used for devices that rely on the initialization of devices
- * initialized as part of the PRE_KERNEL_1 level. These devices cannot use any
- * kernel services during configuration, since they are not yet available.
+ * \li PRE_KERNEL_2: Used for units that rely on the initialization of other
+ * units initialized as part of the PRE_KERNEL_1 level. These units cannot
+ * use any kernel services during configuration, since they are not yet
+ * available.
  * \n
- * \li POST_KERNEL: Used for devices that require kernel services during
+ * \li POST_KERNEL: Used for units that require kernel services during
  * configuration.
  * \n
- * \li POST_KERNEL_SMP: Used for devices that require kernel services during
+ * \li POST_KERNEL_SMP: Used for units that require kernel services during
  * configuration after SMP initialization.
  * \n
  * \li APPLICATION: Used for application components (i.e. non-kernel components)
- * that need automatic configuration. These devices can use all services
+ * that need automatic configuration. These units can use all services
  * provided by the kernel during configuration.
  *
+ * Units are the objects that Zephyr knows how to manage. They represent system
+ * resources that can be initialised and executed based on the level and
+ * priority defined for each unit. Units in some ways are similar to services
+ * or jobs in other init systems and additionally include device drivers.
+ *
  * @param prio The initialization priority of the device, relative to
  * other devices of the same initialization level. Specified as an integer
  * value in the range 0 to 99; lower values indicate earlier initialization.
```

Reply:
We have system services via SYS_INIT() which get invoked from PRE_KERNEL phases, and I fail to see how removing this improves the kernel in a significant way.

Reply:
@anas I did change the doc already, could you check? I did it before you proposal, basically I used "initialization objects" instead of "units".